rpcs3/Emu/Cell/lv2/sys_net/lv2_socket_p2ps.cpp

```diff
@@ -0,0 +1,800 @@
+
+	status      = p2ps_stream_status::stream_listening;
+	max_backlog = backlog;
+
+	return CELL_OK;
+}
+
+std::optional<std::tuple<s32, std::vector<u8>, sys_net_sockaddr>> lv2_socket_p2ps::recvfrom([[maybe_unused]] s32 flags, u32 len, bool is_lock)
+{
+	std::unique_lock<shared_mutex> lock(mutex, std::defer_lock);
+
+	if (is_lock)
+	{
+		lock.lock();
+	}
+
+	if (!data_available)
+	{
+		if (so_nbio)
+		{
+			return {{-SYS_NET_EWOULDBLOCK, {}, {}}};
+		}
+
+		return std::nullopt;
+	}
+
+	const u32 to_give = std::min<u32>(data_available, len);
+	sys_net_sockaddr addr{};
+	std::vector<u8> dest_buf(to_give);
+
+	sys_net.trace("STREAM-P2P socket had %u available, given %u", data_available, to_give);
+
+	u32 left_to_give = to_give;
+	while (left_to_give)
+	{
+		auto& cur_data               = received_data.begin()->second;
+		auto to_give_for_this_packet = std::min(static_cast<u32>(cur_data.size()), left_to_give);
+		memcpy(dest_buf.data() + (to_give - left_to_give), cur_data.data(), to_give_for_this_packet);
+		if (cur_data.size() != to_give_for_this_packet)
+		{
+			auto amount_left = cur_data.size() - to_give_for_this_packet;
+			std::vector<u8> new_vec(amount_left);
+			memcpy(new_vec.data(), cur_data.data() + to_give_for_this_packet, amount_left);
+			auto new_key = (received_data.begin()->first) + to_give_for_this_packet;
+			received_data.emplace(new_key, std::move(new_vec));
+		}
+
+		received_data.erase(received_data.begin());
+
+		left_to_give -= to_give_for_this_packet;
+	}
+
+	data_available -= to_give;
+	data_beg_seq += to_give;
+
+	sys_net_sockaddr_in_p2p* addr_p2p = reinterpret_cast<sys_net_sockaddr_in_p2p*>(&addr);
+	addr_p2p->sin_family              = AF_INET;
+	addr_p2p->sin_addr                = std::bit_cast<be_t<u32>, u32>(op_addr);
+	addr_p2p->sin_port                = op_vport;
+	addr_p2p->sin_vport               = op_port;
+	addr_p2p->sin_len                 = sizeof(sys_net_sockaddr_in_p2p);
+
+	return {{to_give, dest_buf, addr}};
+}
+
+std::optional<s32> lv2_socket_p2ps::sendto([[maybe_unused]] s32 flags, const std::vector<u8>& buf, std::optional<sys_net_sockaddr> opt_sn_addr, bool is_lock)
+{
+	std::unique_lock<shared_mutex> lock(mutex, std::defer_lock);
+
+	if (is_lock)
+	{
+		lock.lock();
+	}
+
+	constexpr u32 max_data_len = (65535 - (sizeof(u16) + sizeof(p2ps_encapsulated_tcp)));
+
+	::sockaddr_in name{};
+	if (opt_sn_addr)
+	{
+		name = sys_net_addr_to_native_addr(*opt_sn_addr);
+	}
+
+	// Prepare address
+	name.sin_family      = AF_INET;
+	name.sin_port        = std::bit_cast<u16, be_t<u16>>(op_port);
+	name.sin_addr.s_addr = op_addr;
+	// Prepares encapsulated tcp
+	p2ps_encapsulated_tcp tcp_header;
+	tcp_header.src_port = vport;
+	tcp_header.dst_port = op_vport;
+	// chop it up
+	std::vector<std::vector<u8>> stream_packets;
+	u32 cur_total_len = buf.size();
+	while (cur_total_len > 0)
+	{
+		u32 cur_data_len;
+		if (cur_total_len >= max_data_len)
+			cur_data_len = max_data_len;
+		else
+			cur_data_len = cur_total_len;
```

Comment:
Seems like a case for min ?
```suggestion
		u32 cur_data_len = std::min(cur_total_len, max_data_len);
```

Reply:
Yep way better :).